File: smart-parking-management/src/api/subscription.ts
import { apiClient } from "@/api/client"
import type { ApiSuccess } from "@/types"

export interface SubscriptionPlan {
  id: number
  name: string
  description: string | null
  per_slot_price: number
  is_active: boolean
  created_at: string | null
  updated_at: string | null
}

export interface Subscription {
  id: number
  parking_owner_id: number
  plan_id: number
  total_slots: number
  total_price: number
  status: string
  created_at: string
  updated_at: string | null
  plan: SubscriptionPlan | null
}

export interface SubscriptionPlanCreate {
  name: string
  description?: string
  per_slot_price: number
  is_active?: boolean
}

export interface SubscriptionPlanUpdate {
  name?: string
  description?: string
  per_slot_price?: number
  is_active?: boolean
}

export interface SubscriptionPurchaseRequest {
  plan_id: number
  total_slots: number
}

export const subscriptionApi = {
  // Subscription Plans (Admin)
  async getPlans(activeOnly: boolean = true) {
    const res = await apiClient.get<ApiSuccess<SubscriptionPlan[]>>("/subscriptions/plans", {
      params: { active_only: activeOnly },
    })
    return res.data.data
  },

  async getPlan(planId: number) {
    const res = await apiClient.get<ApiSuccess<SubscriptionPlan>>(`/subscriptions/plans/${planId}`)
    return res.data.data
  },

  async createPlan(payload: SubscriptionPlanCreate) {
    const res = await apiClient.post<ApiSuccess<SubscriptionPlan>>("/subscriptions/plans", payload)
    return res.data.data
  },

  async updatePlan(planId: number, payload: SubscriptionPlanUpdate) {
    const res = await apiClient.put<ApiSuccess<SubscriptionPlan>>(`/subscriptions/plans/${planId}`, payload)
    return res.data.data
  },

  async deletePlan(planId: number) {
    const res = await apiClient.delete<ApiSuccess<null>>(`/subscriptions/plans/${planId}`)
    return res.data
  },

  // Subscriptions (Owner)
  async getMySubscription() {
    const res = await apiClient.get<ApiSuccess<Subscription | null>>("/subscriptions/me")
    return res.data.data
  },

  async getMySubscriptionStatus() {
    const res = await apiClient.get<ApiSuccess<{ has_subscription: boolean; status: string; message: string; subscription?: Subscription }>>(
      "/subscriptions/me/status",
    )
    return res.data.data
  },

  async purchaseSubscription(payload: SubscriptionPurchaseRequest) {
    const res = await apiClient.post<ApiSuccess<Subscription>>("/subscriptions/purchase", payload)
    return res.data.data
  },

  // Subscriptions (Admin)
  async getSubscription(subscriptionId: number) {
    const res = await apiClient.get<ApiSuccess<Subscription>>(`/subscriptions/${subscriptionId}`)
    return res.data.data
  },

  async updateSubscription(subscriptionId: number, payload: Partial<Subscription>) {
    const res = await apiClient.put<ApiSuccess<Subscription>>(`/subscriptions/${subscriptionId}`, payload)
    return res.data.data
  },

  async cancelSubscription(subscriptionId: number) {
    const res = await apiClient.post<ApiSuccess<Subscription>>(`/subscriptions/${subscriptionId}/cancel`)
    return res.data.data
  },
}

File: smart-parking-api/app/schemas/subscription.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class SubscriptionPlanBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = None
    per_slot_price: float = Field(..., gt=0)
    is_active: bool = True


class SubscriptionPlanCreate(SubscriptionPlanBase):
    pass


class SubscriptionPlanUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = None
    per_slot_price: float | None = Field(None, gt=0)
    is_active: bool | None = None


class SubscriptionPlanOut(SubscriptionPlanBase):
    id: int
    created_at: datetime | None = None
    updated_at: datetime | None = None

    class Config:
        from_attributes = True


class SubscriptionBase(BaseModel):
    plan_id: int
    total_slots: int = Field(..., gt=0)
    total_price: float = Field(..., gt=0)
    status: str = "active"


class SubscriptionCreate(SubscriptionBase):
    parking_owner_id: int


class SubscriptionUpdate(BaseModel):
    status: str | None = None


class SubscriptionOut(SubscriptionBase):
    id: int
    parking_owner_id: int
    created_at: datetime
    updated_at: datetime | None = None
    plan: SubscriptionPlanOut | None = None

    class Config:
        from_attributes = True


class SubscriptionPurchaseRequest(BaseModel):
    plan_id: int
    total_slots: int = Field(..., gt=0)
